### DISCUSSION TITLE
Mixing async/sync Operations in FastAPI Microservices Architecture

### DISCUSSION THREAD

**Discussion by AliYmn** [fastapi/fastapi#12995] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter

router = APIRouter()

@router.get("/items")
async def read_items():
    # Synchronous database call
    items = get_items_from_db()
    return items

def get_items_from_db():
    # Simulate a synchronous database operation
    return [{"item_id": 1, "name": "Item One"}]
```

### Description

Hello everyone,

I'm working on a microservices project using FastAPI and have encountered some challenges related to mixing asynchronous (`async`) and synchronous (`sync`) operations.

![Screenshot 2024-11-27 at 11 23 43](https://github.com/user-attachments/assets/8dc0113f-95d7-465f-97de-2752d717d2e5)
Source: [FastAPI - Async: In a hurry](https://fastapi.tiangolo.com/async/#in-a-hurry)

In our application, we have routers with asynchronous endpoints that call synchronous functions for database access, Redis operations, or external service calls.

My questions are:
1. Performance Impact: How does mixing async and sync code affect the performance of a FastAPI application, especially when scaling in a microservices architecture?
2. Event Loop Blocking: Does calling synchronous functions within async endpoints block the event loop? If so, how significant is this effect under high load?
3. Best Practices: What are the recommended strategies for handling synchronous operations in an asynchronous FastAPI route? Should we use thread pools or asyncio.to_thread to mitigate potential issues?
4. Scalability Concerns: In a horizontally scaled microservices environment, how critical is it to ensure all operations are asynchronous? What are the potential bottlenecks?
5. Synchronous Database: If the database library being used doesn't support asynchronous operations, what is the best approach to handle database calls within asynchronous FastAPI routes? Is it still beneficial to use async endpoints in this case?

I've searched for technical details on this topic but haven't found comprehensive explanations. Can someone provide insights or resources that explain the impact and best practices for mixing async and sync code in FastAPI?

Thank you in advance for your help!

### Operating System

macOS

### FastAPI Version

0.115.5

### Pydantic Version

2.8.0

### Python Version

3.12.3

**Comment by AliYmn** [comment, marked as answer]
I got answers from these links;

1-) https://github.com/Kludex/fastapi-tips?tab=readme-ov-file#2-be-careful-with-non-async-functions
2-) https://thedkpatel.medium.com/fastapi-performance-showdown-sync-vs-async-which-is-better-77188d5b1e3a#:~:text=The%20Async%20Version%20handles%2050.68,winner%20in%20terms%20of%20concurrency
3-) https://youtu.be/QcRq3F8FpSg?si=6CcWFtAVq_Yavr9J&t=987
4-) https://fastapi.tiangolo.com/async/#in-a-hurry
5-) https://github.com/fastapi/fastapi/issues/1066
6-) https://medium.com/@nikhillad01/how-the-event-loop-handles-async-and-sync-requests-in-fastapi-94c292c9376a

**Reply by Kludex** [reply]
I need to write a bit more on that section. But I don't think there are good resources around anyway.

**Reply by AliYmn** [reply]
@Kludex It could be great! but It was tough to find resources about it, but the truth is that using async(without mix) for everything improved performance.
